chainerx_cc/chainerx/backward.cc

```diff
@@ -621,13 +639,29 @@ std::vector<absl::optional<Array>> Grad(
         }
     }
 
-    BackwardImpl{inputs, outputs, actual_backprop_id, double_backprop, std::move(array_node_grad_map)}.Run();
+    // Push initial output grads, Run() assigns to 1 in other case
+    for (size_t i = 0; i < grad_outputs.size(); ++i) {
+        const std::shared_ptr<ArrayBody>& array_body = internal::GetArrayBody(outputs[i]);
+        if (const std::shared_ptr<ArrayNode>& array_node = array_body->GetArrayNode(actual_backprop_id)) {
+            output_grads.emplace_back(grad_outputs[i]);
+            array_node_grad_map.emplace(array_node.get(), internal::GradRef{&output_grads.back()});
+        }
+    }
 
-    // input_grads may contain unmodified array bodies (nullptr) for arrays that are not included in the graph.
-    // Those grads are returned as nullopt.
-    for (absl::optional<Array>& grad : input_grads) {
-        if (grad.has_value() && internal::GetArrayBody(*grad) == nullptr) {
-            grad = absl::nullopt;
+    BackwardImpl{inputs, outputs, actual_backprop_id, double_backprop, std::move(array_node_grad_map), retain_grad}.Run();
+
+    for (size_t i = 0; i < input_grads.size(); ++i) {
+        absl::optional<Array>& input_grad = input_grads[i];
+        if (input_grad.has_value()) {
+            // If the allocated input_grad was not used during calculation due to a not connected graph
+            // the body was never initialized and it remains as null
+            if (internal::GetArrayBody(*input_grad) == nullptr) {
+                input_grad = absl::nullopt;
+                // Explicitly assign the grad to the input array if specified
+                // Input grads are not set by default during Backward run
```

Comment:
It seems this comment should be placed in the next clause?
How about moving the previous one to this clause, too?